internal/concierge/impersonator/impersonator.go

```diff
@@ -0,0 +1,225 @@
+	log         logr.Logger
+}
+
+func New(cache *authncache.Cache, jsonDecoder runtime.Decoder, log logr.Logger) (http.Handler, error) {
+	return newInternal(cache, jsonDecoder, log, func() (*rest.Config, error) {
+		client, err := kubeclient.New()
+		if err != nil {
+			return nil, err
+		}
+		return client.JSONConfig, nil
+	})
+}
+
+func newInternal(cache *authncache.Cache, jsonDecoder runtime.Decoder, log logr.Logger, getConfig func() (*rest.Config, error)) (*proxy, error) {
+	kubeconfig, err := getConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster config: %w", err)
+	}
+
+	serverURL, err := url.Parse(kubeconfig.Host)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse host URL from in-cluster config: %w", err)
+	}
+
+	kubeTransportConfig, err := kubeconfig.TransportConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport config: %w", err)
+	}
+	kubeTransportConfig.TLS.NextProtos = []string{"http/1.1"} // TODO huh?
+
+	kubeRoundTripper, err := transport.New(kubeTransportConfig)
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport: %w", err)
+	}
+
+	reverseProxy := httputil.NewSingleHostReverseProxy(serverURL)
+	reverseProxy.Transport = kubeRoundTripper
+	reverseProxy.FlushInterval = 200 * time.Millisecond // the "watch" verb will not work without this line
+
+	return &proxy{
+		cache:       cache,
+		jsonDecoder: jsonDecoder,
+		proxy:       reverseProxy,
+		log:         log,
+	}, nil
+}
+
+func (p *proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log := p.log.WithValues(
+		"url", r.URL.String(),
+		"method", r.Method,
+	)
+
+	if err := ensureNoImpersonationHeaders(r); err != nil {
+		log.Error(err, "impersonation header already exists")
+		http.Error(w, "impersonation header already exists", http.StatusBadRequest)
+		return
+	}
+
+	// Never mutate request (see http.Handler docs).
+	newR := r.Clone(r.Context())
+
+	authentication, authenticated, err := bearertoken.New(authenticator.TokenFunc(func(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+		tokenCredentialReq, err := extractToken(token, p.jsonDecoder)
+		if err != nil {
+			log.Error(err, "invalid token encoding")
+			return nil, false, &httpError{message: "invalid token encoding", code: http.StatusBadRequest}
+		}
+
+		log = log.WithValues(
+			"authenticator", tokenCredentialReq.Spec.Authenticator,
+		)
+
+		userInfo, err := p.cache.AuthenticateTokenCredentialRequest(newR.Context(), tokenCredentialReq)
+		if err != nil {
+			log.Error(err, "received invalid token")
+			return nil, false, &httpError{message: "invalid token", code: http.StatusUnauthorized}
+		}
+		if userInfo == nil {
+			log.Info("received token that did not authenticate")
+			return nil, false, &httpError{message: "not authenticated", code: http.StatusUnauthorized}
+		}
+		log = log.WithValues("userID", userInfo.GetUID())
+
+		return &authenticator.Response{User: userInfo}, true, nil
```

Comment:
We will probably care about audiences when we add support for SA tokens.

Reply:
Resolving this since it's likely going away in the refactor.